### PR TITLE
Null runner

### PR DESCRIPTION
**Pull request by erikh**
**PLEASE NOTE**: #13 needs to be cleared before this as it includes it.

This is a null runner which we will use for further exercising tinyCI's agents in integration testing.

### CODE REVIEW

File: fw/signal/signal.go [hunk 1]
```diff
@@ -57,8 +58,8 @@ func (ctx *Context) HandleCancel() {
 			}
 		}
 
-		fmt.Println("Signal received; will wait 10 seconds for cleanup to occur")
-		time.Sleep(10 * time.Second)
+		fmt.Printf("Signal received; will wait %v for cleanup to occur\n", waitTime)
+		time.Sleep(waitTime)
```
Comment (capoferro) [inline fw/signal/signal.go]: Is there any way to make this check for success and not wait the whole time?

File: runners/null-runner/runner.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+// Runner encapsulates an infinite lifecycle overlay-runner.
+type Runner struct {
+	Config    *config.Config
+	NextState bool
+}
+
+// Init is the bootstrap of the runner.
+func (r *Runner) Init(ctx *fw.Context) *errors.Error {
+	rand.Seed(time.Now().UnixNano())
```
Comment (capoferro) [inline runners/null-runner/runner.go]: Why `UnixNano` and not just `Unix`? Not a problem or at all important, just curious.

File: runners/overlay-runner/config/config.go [hunk 3]
```diff
@@ -1,68 +1,22 @@
...
 import (
-	"os"
-	"path"
-
-	"github.com/tinyci/ci-agents/clients/asset"
-	"github.com/tinyci/ci-agents/clients/log"
-	"github.com/tinyci/ci-agents/clients/queue"
-	"github.com/tinyci/ci-agents/config"
-	"github.com/tinyci/ci-runners/git"
+	fwConfig "github.com/tinyci/ci-runners/fw/config"
```
Comment (capoferro) [inline runners/overlay-runner/config/config.go]: this no longer conflicts with `ci-agents/config`. Should this still have an overridden import name?

File: README.md [hunk 4]
```diff
@@ -21,6 +21,12 @@ container can write all it wants to the git repository's directory, but the
 overlayfs will capture that. At the end of the run, the overlayfs is removed
 from disk, returning the repository to a pristine state.
 
+## Framework
+
+We have a runner framework to make it easy to build runners; please see our
+[GoDoc](https://godoc.org/github.com/tinyci/ci-runners/fw) for more information
```
[the remark this answers is not in the document]
Comment (erikh) [reply README.md]: Correct

File: cmd/overlay-runner/main.go [hunk 5]
```diff
@@ -1,164 +1,24 @@
...
-		utils.ErrOut(err)
-	}
-}
-
-func loop(ctx *cli.Context) error {
-	// we reload the clients on each run
-	c, err := config.Load(ctx.GlobalString("config"))
+`,
+		Launch:       &runner.Runner{},
+		TeardownTime: 10 * time.Second,
```
[the remark this answers is not in the document]
Comment (erikh) [reply cmd/overlay-runner/main.go]: Ok, will change.

File: fw/fw.go [hunk 6]
```diff
@@ -0,0 +1,224 @@
+// Package fw is composed of a framework for designing tinyCI runner agents.
+//
+// Inside are several packages that cover config files, signals, and other
+// features. You will want to read the docs for them too.
```
[the remark this answers is not in the document]
Comment (erikh) [reply fw/fw.go]: They are linked as a part of the standard godoc process, it's similar to rdoc in how things are assembled as a single unit of documentation.

File: fw/fw.go [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+type Runner interface {
+
+	//
+	// Lifecycle hooks
+	//
+	// Init is the entrypoint of the runner application and will be run shortly
+	// after command line arguments are processed.
+	Init(*Context) *errors.Error
+	// BeforeNextRun is executed to set up the run but not actually execute it.
+	BeforeNextRun(*Context) *errors.Error
```
[the remark this answers is not in the document]
Comment (erikh) [reply fw/fw.go]: OK, I'll change it to `BeforeRun()` -- I want it to be clear in the event we need more hooks later.

And htat kinda-sorta answers your other question; if/when we need After(), we'll add it. Until then it's architecture astronomy, but I totally get why you would ask.

File: fw/signal/signal.go (hunk 1 above)
Comment (erikh) [reply fw/signal/signal.go]: Um there are some tricks we can do with the context to do that; but for now this works and is less than ideal but something that we can use without too much in the heisenbug department.

I'd like to keep it this way for now and maybe make it smoother later, by allowing the user to close a channel to signal when they are done shutting down properly.

Comment (erikh) [reply fw/signal/signal.go]: Let me know if you're comfortable with that.

File: runners/null-runner/runner.go (hunk 2 above)
Comment (erikh) [reply runners/null-runner/runner.go]: UnixNano(): https://godoc.org/time#Time.UnixNano

Unix has a granularity of one second, UnixNano is harder to repeat a seed for in a quick restart loop.

File: runners/overlay-runner/config/config.go (hunk 3 above)
Comment (erikh) [reply runners/overlay-runner/config/config.go]: Hm. I'll peek deeper in a minute but I'm happy to change it too; I probably had both in there at some point.

**Comment by erikh** [comment]
Taking patches to #13 only, please comment there. Will close this for now.

**Pull request closed without merging**